File: clients/swift/Sources/KeySync/LinuxKeychain.swift
import Foundation

/// Linux keychain access via libsecret's `secret-tool` CLI.
///
/// Falls back to the `secret-tool` command-line tool, which is part of
/// libsecret. Requires `libsecret-tools` to be installed and a running
/// secret service (GNOME Keyring, KDE Wallet, KeePassXC).
struct LinuxKeychain {

    private let secretToolPath = "/usr/bin/secret-tool"

    func getSecret(service: String, account: String) throws -> String {
        let output = try runSecretTool(arguments: ["lookup", "service", service, "account", account])
        let value = output.trimmingCharacters(in: .whitespacesAndNewlines)
        if value.isEmpty {
            throw KeySyncError.notFound
        }
        return value
    }

    func listSecrets() throws -> [(service: String, account: String)] {
        // secret-tool search outputs lines like:
        //   service = keysync/global
        //   account = MY_KEY
        //   password = <value>
        // with blank lines between entries.
        let output = try runSecretTool(arguments: ["search", "service", "keysync"])
        let lines = output.components(separatedBy: .newlines)

        var results: [(service: String, account: String)] = []
        var currentService: String?
        var currentAccount: String?

        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty {
                // Blank line = end of entry
                if let svc = currentService, let acct = currentAccount {
                    results.append((svc, acct))
                }
                currentService = nil
                currentAccount = nil
                continue
            }
            if trimmed.hasPrefix("service") {
                currentService = parseAttributeValue(trimmed)
            } else if trimmed.hasPrefix("account") {
                currentAccount = parseAttributeValue(trimmed)
            }
        }
        // Last entry if no trailing blank line
        if let svc = currentService, let acct = currentAccount {
            results.append((svc, acct))
        }

        return results
    }

    // MARK: - Private

    private func runSecretTool(arguments: [String]) throws -> String {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: secretToolPath)
        process.arguments = arguments

        let outputPipe = Pipe()
        let errorPipe = Pipe()
        process.standardOutput = outputPipe
        process.standardError = errorPipe

        try process.run()
        process.waitUntilExit()

        let outputData = outputPipe.fileHandleForReading.readDataToEndOfFile()
        let errorData = errorPipe.fileHandleForReading.readDataToEndOfFile()

        if process.terminationStatus != 0 {
            let errorOutput = String(data: errorData, encoding: .utf8) ?? ""
            // secret-tool returns exit code 1 when the secret is not found
            if errorOutput.contains("not found") || errorOutput.isEmpty {
                throw KeySyncError.notFound
            }
            throw KeySyncError.keychainError("secret-tool exited with status \(process.terminationStatus): \(errorOutput)")
        }

        return String(data: outputData, encoding: .utf8) ?? ""
    }

    private func parseAttributeValue(_ line: String) -> String {
        let parts = line.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
        guard parts.count == 2 else { return "" }
        return String(parts[1]).trimmingCharacters(in: .whitespaces)
    }
}

File: clients/java/src/main/java/io/keysync/LinuxKeychain.java
package io.keysync;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Linux keychain access via libsecret's {@code secret-tool} CLI.
 *
 * <p>Requires the {@code libsecret-tools} package to be installed and a
 * running secret service (GNOME Keyring, KDE Wallet, KeePassXC, etc.).
 */
class LinuxKeychain implements KeychainProvider {

    @Override
    public String getSecret(String service, String account) throws KeySyncException {
        ProcessBuilder pb = new ProcessBuilder(
                "secret-tool", "lookup",
                "keysync-service", service,
                "keysync-key", account
        );
        pb.redirectErrorStream(false);

        try {
            Process process = pb.start();
            String stdout = readString(process.getInputStream());
            String stderr = readString(process.getErrorStream());
            int exitCode = process.waitFor();

            if (exitCode == 1) {
                // secret-tool returns 1 when the secret is not found
                throw new KeySyncException(
                        KeySyncError.NOT_FOUND,
                        "secret not found: " + service + "/" + account
                );
            }

            if (exitCode != 0) {
                throw new KeySyncException(
                        KeySyncError.KEYCHAIN_ERROR,
                        "secret-tool lookup failed with code " + exitCode
                            + ": " + stderr.trim()
                );
            }

            String value = stdout.trim();
            if (value.isEmpty()) {
                throw new KeySyncException(
                        KeySyncError.NOT_FOUND,
                        "secret not found: " + service + "/" + account
                );
            }
            return value;

        } catch (IOException e) {
            throw new KeySyncException(
                    KeySyncError.KEYCHAIN_ERROR,
                    "failed to run secret-tool: " + e.getMessage()
                        + " (is libsecret-tools installed?)",
                    e
            );
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new KeySyncException(
                    KeySyncError.KEYCHAIN_ERROR,
                    "interrupted while waiting for secret-tool",
                    e
            );
        }
    }

    @Override
    public List<Credential> listSecrets() throws KeySyncException {
        ProcessBuilder pb = new ProcessBuilder("secret-tool", "search", "keysync-service", "keysync");
        pb.redirectErrorStream(true);

        try {
            Process process = pb.start();
            String output = readString(process.getInputStream());
            int exitCode = process.waitFor();

            if (exitCode != 0) {
                // secret-tool not available or no results
                return Collections.emptyList();
            }

            List<Credential> results = new ArrayList<>();
            String currentService = null;
            String currentAccount = null;

            for (String line : output.split("\n")) {
                line = line.trim();
                if (line.isEmpty()) {
                    // Blank line = end of entry
                    if (currentService != null && currentAccount != null) {
                        results.add(new Credential(currentService, currentAccount));
                    }
                    currentService = null;
                    currentAccount = null;
                    continue;
                }

                if (line.startsWith("keysync-service")) {
                    currentService = parseValue(line);
                } else if (line.startsWith("keysync-key")) {
                    currentAccount = parseValue(line);
                }
            }

            // Handle last entry if no trailing blank line
            if (currentService != null && currentAccount != null) {
                results.add(new Credential(currentService, currentAccount));
            }

            return results;

        } catch (IOException e) {
            throw new KeySyncException(
                    KeySyncError.KEYCHAIN_ERROR,
                    "failed to run secret-tool search: " + e.getMessage(),
                    e
            );
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new KeySyncException(
                    KeySyncError.KEYCHAIN_ERROR,
                    "interrupted while searching keychain",
                    e
            );
        }
    }

    // --- helpers ---

    private static String readString(java.io.InputStream stream) throws IOException {
        StringBuilder sb = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (sb.length() > 0) {
                    sb.append('\n');
                }
                sb.append(line);
            }
        }
        return sb.toString();
    }

    /**
     * Parses "keysync-service = keysync/global" and returns "keysync/global".
     */
    private static String parseValue(String line) {
        int eqIdx = line.indexOf('=');
        if (eqIdx < 0) {
            return "";
        }
        return line.substring(eqIdx + 1).trim();
    }
}
